### PR TITLE
Display changes

### LINKED ISSUE #73: Remove set_plot_options
…or deprecate and hide it from the docs.

R has a notion of the “default graphics device”, i.e. `getOption('device')`, which is either a string containing a function name or a function.

we could leverage this in two ways:

1. use it with custom options:

	```r
	dev <- getOption('device')
	if (!is.function(dev)) {
		dev <- get(dev)
	}
	do.call(dev, as.list(options('jupyter.plot')))
	```

2. require the user to override the defaults by setting it to a function:

	```r
	#our initialization
	options(device = function(file.name, width = 7, height = 7,...) {
		png(file.name, width = width, height = height, ...)
	})
	
	…
	
	dev <- getOption('device')
	if (!is.function(dev)) {
		dev <- get(dev)
	}
	dev()
	```

---

the only problem: we want to allow different mime types (PNG, JPG, SVG) for plots: if it’s a function, we can’t derive the type from the option.

### PR DESCRIPTION
**Pull request by flying-sheep**
i think this should be the way everyone’s happy with.

depends on ~~IRkernel/IRdisplay#3’s function `display_alternatives` and~~ [IRkernel/repr](https://github.com/IRkernel/repr) (in order to display multiple alternative formats of one output, and to create those formats, respectively)

sorry for the mess of commits, i brought the notebooks up to date too. relevant changes should start [here](https://github.com/IRkernel/IRkernel/pull/82/files#diff-6eae7f9cccd978b5b4d819baa7a34b1e)

this fixes #73 and lays foundations for #59 (by depending on repr)

### CODE REVIEW

File: R/execution.r [hunk 1]
```diff
@@ -83,12 +62,13 @@ execute = function(request) {
...
-      replayPlot(plotobj)
-      dev.off()
-      display_png(file=tf)
+  send_plot <- function(plotobj) {
+      params <- list()
+      for (mime in getOption('jupyter.plot_mimetypes')) {
+          r <- mime2repr[[mime]](plotobj)
+          params[[mime]] <- if (is.raw(r)) base64encode(r) else r
+      }
+      do.call(display_alternatives, params)
```
Comment (JanSchulz) [inline R/execution.r]: Ok, I was going to argue that this should send a `execution_reply` but in python this also uses the `display` function :-/

But this then means that the kernel depends on the display package (due to the useage in sending plots) and the display package depends on the kernel (due to the actual sending of the `display_data` reply). I'm not sure if that's a problem in R.

Comment (takluyver) [inline R/execution.r]: This is already the case - IRdisplay provides the display API, that I also want to make usable from the R magic in a Python notebook, so the same code works in both. However, it doesn't know how to actually display the data - it needs to be run by something that gives it a base display function to send the data. IRkernel does this. It's not very nice, but it seems to work.

I'm still not convinced that display_alternatives is necessary, though.

**Comment by flying-sheep** [comment]
so regarding #83:

you call that a hack? [1009088](https://github.com/IRkernel/IRkernel/commit/10090886efb30cc4948772b7e3b61ba9cfae4d7e) is a hack!

A hack

**Comment by flying-sheep** [comment]
btw. i think that this is now independent from IRkernel/IRdisplay#3, right?

so we can pull this in and bikeshed over the IRdisplay apis the following weeks ;)

**Comment by takluyver** [comment]
Yes, I think so - I tested it without that PR (using #83).

I also want to think a bit about repr APIs, though - I'm a bit confused as to why repr has a class, rather than just a set of methods returning string/raw vectors.

**Comment by flying-sheep** [comment]
> I also want to think a bit about repr APIs, though - I'm a bit confused as to why repr has a class, rather than just a set of methods returning string/raw vectors.

let’s discuss this in IRkernel/repr#1

**Pull request merged**